doctr/models/recognition/master.py

```diff
@@ -197,56 +212,119 @@ def __init__(
+            model_output: predicted logits of the model
+            seq_len: lengths of each gt word inside the batch
+
+        Returns:
+            The loss of the model on the batch
+        """
+        # Input length : number of timesteps
+        input_len = tf.shape(model_output)[1]
+        # Add one for additional <eos> token
+        seq_len = seq_len + 1
+        # One-hot gt labels
+        oh_gt = tf.one_hot(gt, depth=model_output.shape[2])
+        # Compute loss
+        cce = tf.nn.softmax_cross_entropy_with_logits(oh_gt, model_output)
+        # Compute mask
+        mask_values = tf.zeros_like(cce)
+        mask_2d = tf.sequence_mask(seq_len, input_len)
+        masked_loss = tf.where(mask_2d, cce, mask_values)
+        ce_loss = tf.math.divide(tf.reduce_sum(masked_loss, axis=1), tf.cast(seq_len, tf.float32))
+
+        return tf.expand_dims(ce_loss, axis=1)
+
+    def call(
+        self,
+        inputs: tf.Tensor,
+        labels: Optional[List[str]] = None,
+        return_model_output: bool = False,
+        return_preds: bool = False,
+        **kwargs
+    ) -> Dict[str, Any]:
         """Call function for training
 
         Args:
             inputs: images
-            labels: tensor of labels
+            labels: list of str labels
+            return_model_output: if True, return logits
+            return_preds: if True, decode logits
 
         Return:
-            Computed logits
+            A dictionnary containing eventually loss, logits and predictions.
         """
+
         # Encode
         feature = self.feature_extractor(inputs, **kwargs)
         b, h, w, c = (tf.shape(feature)[i] for i in range(4))
         feature = tf.reshape(feature, shape=(b, h * w, c))
         encoded = feature + self.feature_pe[:, :h * w, :]
 
-        tgt_mask = self.make_mask(labels)
+        out: Dict[str, tf.Tensor] = {}
+
+        if labels is not None:
+            # Compute target: tensor of gts and sequence lengths
+            gt, seq_len = self.compute_target(labels)
+
+            tgt_mask = self.make_mask(gt)
+
+            # Compute logits
+            output = self.decoder(gt, encoded, tgt_mask, None, training=True)
+            logits = self.linear(output)
+            if return_model_output:
+                out['out_map'] = logits
+
+            # Compute loss
+            out['loss'] = self.compute_loss(logits, gt, seq_len)
 
-        output = self.decoder(labels, encoded, tgt_mask, None, training=True)
-        logits = self.linear(output)
+            if return_preds:
+                predictions = self.postprocessor(logits)
+                out['preds'] = predictions
 
-        return logits
+        else:
+            raw_predictions, logits = self.decode(encoded)
+            if return_model_output:
+                out['out_map'] = logits
+            if return_preds:
+                predictions = self.postprocessor(logits)
+                out['preds'] = predictions
+
+        return out
```

Comment:
Could you rewrite that so that you get the same structure as in CRNN or SAR?
Most conditions should be independent of the others